### PR TITLE
Delete buggy `stat(::Integer)` method

### LINKED ISSUE #51710: ispath returning false positive result for the given argument
julia> ispath("/tmp")
true

julia> ispath("1")
false

julia> ispath(1)
true

image

### PR DESCRIPTION
**Pull request by LilithHafner**
"Where did someone get a RawFD as an integer anyway?" -@stefankarpinski

See also #51711

Fixes #51710

### CODE REVIEW

**Comment by jakobnissen** [comment]
This is specifically tested for, hence the tests fail. What did triage agree to yesterday? I can modify https://github.com/JuliaLang/julia/pull/51711 accordingly.
Would a deprecation for a (couple of) releases make sense?

**Comment by LilithHafner** [comment]
I opened this PR while screen sharing during triage, but it wasn't explicitly a triage decision. I just want to see what nanosoldier has to say about this.

@nanosoldier `runtests()`

I think the problem with `ispath(6)` is that `stat(6)` works, not that `ispath` is duck-typed.

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/54855#issuecomment-2180686965) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/a1dd036_vs_a14cc38/report.html) is available.

**Comment by LilithHafner** [comment]
> Where did someone get a RawFD as an integer anyway?

```julia
julia> fd(open("tmp", "w"))
23
```

**Comment by LilithHafner** [comment]
The docs don't indicate that `stat` should accept an integer:

```
  stat(file)

Return a structure whose fields contain information about the file. The fields of the structure are:

...
```

**Comment by LilithHafner** [comment]
Of the 19 packages that failed nano-soldier only on this version, 14 were false positives, 3 were due to `stat(::IOStream)` failing (fixed in 6b29410) and only 2 require trivial downstream fixes (https://github.com/halleysfifthinc/C3D.jl/pull/19, https://github.com/emmt/InterProcessCommunication.jl/pull/10).

There is perhaps still hope for this in 1.x

**Comment by halleysfifthinc** [comment]
> The docs don't indicate that `stat` should accept an integer:

Actually, the docs don't indicate at all what  `stat` accepts. I'm happy to accept @LilithHafner 's PR, but that doesn't feel like the complete solution IMO. 

- Could we update the `fd` function to return a `RawFD`? IIRC, changing return types isn't considered a breaking change. Has that been considered and/or is there a reason not to?
- The `stat` docstring should be updated (as part of this PR imo), because the lack of clarity around the proper/desired arguments for `stat` might be why code like mine (i.e. `stat(fd(io))`) exists. A docstring should be clear enough that users don't need to call `methods` to understand what argument types it accepts.
  - A quick `methods(stat)` and blame shows that a `stat(::IOStream) = stat(fd(io))` method has existed for much longer than my code, which makes me suspect that the lack of clarity in the `stat` docstring lead me to the unnecessary indirection through `fd` instead of a direct `stat(io)` call.
  - Additional explanation for the `fd` function and `RawFD` type docstrings would also be helpful. As it stands currently, they appear to be completely redundant, and its not clear why and when you should use one or the other.

**Comment by IanButterworth** [comment]
Just came across this.
```
julia> isdir(false)
false

julia> isfile(false)
false
```

**Comment by LilithHafner** [comment]
What did you expect? `false` is neither a file nor a directory. :smile:

**Comment by vtjnash** [comment]
False is a file for me lol
```
$ echo "isfile(false)" > isfilefalse.jl
$ <isfilefalse.jl ./julia
true
```

**Review by fingolfin — APPROVED** [review]
Looks good to me esp. once #55080 is merged?

File: base/iostream.jl [hunk 1]
```diff
@@ -54,7 +54,7 @@ to synchronous `File`'s and `IOStream`'s not to any of the asynchronous streams.
 """
 fd(s::IOStream) = Int(ccall(:jl_ios_fd, Clong, (Ptr{Cvoid},), s.ios))
 
-stat(s::IOStream) = stat(fd(s))
+stat(s::IOStream) = stat(RawFD(fd(s)))
```
[the remark this answers is not in the document]
Comment (LilithHafner) [reply base/iostream.jl:57]:
```suggestion
stat(s::IOStream) = stat(fd(s))
```

#55080 is merged.

**Comment by LilithHafner** [comment]
@nanosoldier `runtests()`

**Comment by mbauman** [comment]
Why not a deprecation?

**Comment by LilithHafner** [comment]
A deprecation would not fix #51710 and I'm optimistic that, with #55080, this is practically non-breaking which makes it plausible to remove outright.

**Comment by mbauman** [comment]
I suppose I'm just more conservative here — to me this isn't worth an immediate minor change when the deprecation is _so straightforward_.  It's great that you even already tackled the PkgEval failures, but I see that as rescuing two canaries from a particularly deep and obscure branch of our coal mine.  I don't know who else is down there, and we might as well sound the alarm for anyone that's listening.

**Comment by LilithHafner** [comment]
Two of the original PkgEval failures are fixed by #55080. But you are right that there was one that needed to be fixed downstream. We can deprecate.

**Comment by LilithHafner** [comment]
This is still worth removing eventually because of #51710. I tagged 1.14 to be conservative, but would be open to landing it in 1.13, too.

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/54855#issuecomment-2250392041) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/cce422a_vs_bcd9cce/report.html) is available.

**Comment by giordano** [comment]
Failure in [`InterProcessCommunication`](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/cce422a_vs_bcd9cce/InterProcessCommunication.primary.log) looks real (CC: @emmt), others look noise at a quick glance.

**Comment by LilithHafner** [comment]
I agree on both counts. That failure is fixed in https://github.com/emmt/InterProcessCommunication.jl/pull/10.

**Comment by LilithHafner** [comment]
From triage: Keno and Oscar and Lilith are fine with removing this in 1.12 (and postponing the removal if anyone notices)

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/54855#issuecomment-2317606508) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/cf8fbdd_vs_cebfd7b/report.html) is available.

**Comment by oscardssmith** [comment]
Let's try a merge and see if anything breaks.

**Pull request merged**